// testcafe/lib/role/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const nanoid_1 = __importDefault(require("nanoid"));
const phase_1 = __importDefault(require("./phase"));
const type_assertions_1 = require("../errors/runtime/type-assertions");
const wrap_test_function_1 = __importDefault(require("../api/wrap-test-function"));
const test_page_url_1 = require("../api/test-page-url");
const actions_1 = require("../test-run/commands/actions");
const marker_symbol_1 = __importDefault(require("./marker-symbol"));
const testcafe_hammerhead_1 = require("testcafe-hammerhead");
class Role extends events_1.EventEmitter {
    constructor(loginPage, initFn, options = {}) {
        super();
        this[marker_symbol_1.default] = true;
        this.id = nanoid_1.default(7);
        this.phase = loginPage ? phase_1.default.uninitialized : phase_1.default.initialized;
        this.loginPage = loginPage;
        this.initFn = initFn;
        this.opts = options;
        this.url = null;
        this.stateSnapshot = testcafe_hammerhead_1.StateSnapshot.empty();
        this.initErr = null;
    }
    async _navigateToLoginPage(testRun) {
        const navigateCommand = new actions_1.NavigateToCommand({
            url: this.loginPage,
            forceReload: true
        });
        await testRun.executeCommand(navigateCommand);
    }
    async _storeStateSnapshot(testRun) {
        if (this.initErr)
            return;
        this.stateSnapshot = await testRun.getStateSnapshot();
    }
    async _executeInitFn(testRun) {
        try {
            testRun.disableDebugBreakpoints = false;
            await this.initFn(testRun);
        }
        catch (err) {
            this.initErr = err;
        }
        finally {
            testRun.disableDebugBreakpoints = true;
        }
    }
    async initialize(testRun) {
        this.phase = phase_1.default.pendingInitialization;
        await testRun.switchToCleanRun();
        await this._navigateToLoginPage(testRun);
        await this._executeInitFn(testRun);
        await this._storeStateSnapshot(testRun);
        if (this.opts.preserveUrl)
            this.url = await testRun.getCurrentUrl();
        this.phase = phase_1.default.initialized;
        this.emit('initialized');
    }
}
function createRole(loginPage, initFn, options = {}) {
    type_assertions_1.assertType(type_assertions_1.is.string, 'Role', '"loginPage" argument', loginPage);
    type_assertions_1.assertType(type_assertions_1.is.function, 'Role', '"initFn" argument', initFn);
    type_assertions_1.assertType(type_assertions_1.is.nonNullObject, 'Role', '"options" argument', options);
    if (options.preserveUrl !== void 0)
        type_assertions_1.assertType(type_assertions_1.is.boolean, 'Role', '"preserveUrl" option', options.preserveUrl);
    loginPage = test_page_url_1.resolvePageUrl(loginPage);
    initFn = wrap_test_function_1.default(initFn);
    return new Role(loginPage, initFn, options);
}
exports.createRole = createRole;
function createAnonymousRole() {
    return new Role(null, null);
}
exports.createAnonymousRole = createAnonymousRole;
